relayer/pkg/chainlink/txm/txm.go

```diff
@@ -69,135 +81,167 @@ func (txm *starktxm) Name() string {
-
-			// async process of tx batches
-			var wg sync.WaitGroup
-			wg.Add(len(txsByAccount))
-			for key, txs := range txsByAccount {
-				txm.lggr.Debugw("batch details", "accountAddress", key.accountAddress, "senderAddress", key.senderAddress, "txs", txs)
-				go func(senderAddress caigotypes.Hash, accountAddress caigotypes.Hash, txs []caigotypes.FunctionCall) {
-					// fetch key matching sender address
-					key, err := txm.ks.Get(senderAddress.String())
-					if err != nil {
-						txm.lggr.Errorw("failed to retrieve key", "id", senderAddress, "error", err)
-					} else {
-						// parse key to expected format
-						privKeyBytes := key.Raw()
-						privKey := caigotypes.BigToHex(caigotypes.BytesToBig(privKeyBytes))
-
-						// broadcast batch based on account address
-						hash, err := txm.broadcastBatch(ctx, privKey, accountAddress, txs)
-						if err != nil {
-							txm.lggr.Errorw("transaction failed to broadcast", "error", err, "account", accountAddress, "batchTx", txs)
-						} else {
-							txm.lggr.Infow("transaction broadcast", "txhash", hash)
-						}
-					}
-					wg.Done()
-				}(key.senderAddress, key.accountAddress, txs)
+			// parse key to expected format
+			privKeyBytes := key.Raw()
+			privKey := caigotypes.BigToHex(caigotypes.BytesToBig(privKeyBytes))
+
+			// broadcast tx serially - wait until accepted by mempool before processing next
+			hash, err := txm.broadcast(ctx, privKey, tx.accountAddress, tx.call)
+			if err != nil {
+				txm.lggr.Errorw("transaction failed to broadcast", "error", err, "tx", tx.call)
+			} else {
+				txm.lggr.Infow("transaction broadcast", "txhash", hash)
 			}
-			wg.Wait()
-			tick = time.After(utils.WithJitter(txm.cfg.TxSendFrequency()) - time.Since(start))
-		case <-txm.stop:
-			return
 		}
 	}
 }
 
 const FEE_MARGIN uint64 = 115
 
-func (txm *starktxm) broadcastBatch(ctx context.Context, privKey string, accountAddress caigotypes.Hash, txs []caigotypes.FunctionCall) (txhash string, err error) {
+func (txm *starktxm) broadcast(ctx context.Context, privKey string, accountAddress caigotypes.Hash, tx caigotypes.FunctionCall) (txhash string, err error) {
+	txs := []caigotypes.FunctionCall{tx}
+	client, err := txm.client.Get()
+	if err != nil {
+		return txhash, fmt.Errorf("broadcast: failed to fetch client: %+w", err)
+	}
 	// create new account
-	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), txm.client.Gw, caigo.AccountVersion1)
+	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), client.Gw, caigo.AccountVersion1)
+	if err != nil {
+		return txhash, fmt.Errorf("failed to create new account: %+w", err)
+	}
+
+	nonce, err := txm.nonce.NextSequence(accountAddress, client.Gw.ChainId)
 	if err != nil {
-		return txhash, errors.Errorf("failed to create new account: %s", err)
+		return txhash, fmt.Errorf("failed to get nonce: %+w", err)
 	}
 
 	// get fee for txm
+	// optional - pass nonce to fee estimate (if nonce gets ahead, estimate may fail)
```

Comment:
Actually, if you don't pass in a nonce, it'll fetch one from the chain for you anyway which is what you were trying to avoid: https://github.com/smartcontractkit/caigo/blob/41120ca1f9f3689da24469fe19942707abcf1425/account.go#L261-L266

Reply:
that's true! but that's not a reliable source for the nonce - especially if you need to send multiple transactions soon after each other since the chain may not have included the first tx, you would end broadcasting two txs that have the same nonce.

for estimation this is fine since we are just looking for fee values, but for actual broadcasting we'd run into transaction failures.